utf8proc.c

```diff
@@ -822,3 +822,289 @@ UTF8PROC_DLLEXPORT utf8proc_uint8_t *utf8proc_NFKC_Casefold(const utf8proc_uint8
+       decomposed string in the worst case. since we only need to compare
+       them as-if they are normalized, we can just go one combining class
+       at a time. we have to be careful around ends of strings to make
+       sure the string views are properly updated to NOT FURTHER THAN the
+       first difference in the strings, which may be a large combining seq.
+       */
+    while (!a_consume && !b_consume) {
+      /* do we need to decompose more? */
+      if (decomposing_current.a_decomposed_pos >= decomposing_current.a_decomposed_len) {
+        a_consume = true;
+        a->str.ptr += a_consumed;
+        a->str.len -= a_consumed * a_len_terminated;
+      }
+      if (decomposing_current.b_decomposed_pos >= decomposing_current.b_decomposed_len) {
+        b_consume = true;
+        b->str.ptr += b_consumed;
+        b->str.len -= b_consumed * b_len_terminated;
+      }
+      if (a_consume || b_consume) {
+        continue;
+      }
+      /* get the combining class of each current code point */
+      a_combining_class = unsafe_get_property(decomposing_current.a_decomposed[decomposing_current.a_decomposed_pos])->combining_class;
+      b_combining_class = unsafe_get_property(decomposing_current.b_decomposed[decomposing_current.b_decomposed_pos])->combining_class;
+      /* static analysis guards, always false in practice */
+      if (a_combining_class/CHAR_BIT >= combining_classes_finished_len) {
+        a->error = UTF8PROC_ERROR_OVERFLOW;
+        a->str_at_error = a->str;
+      }
+      if (b_combining_class/CHAR_BIT >= combining_classes_finished_len) {
+        b->error = UTF8PROC_ERROR_OVERFLOW;
+        b->str_at_error = b->str;
+      }
+      if (a->error || b->error) {
+        if (combining_initialized) {
+          a->str = a_combining_start;
+          b->str = b_combining_start;
+        }
+        return;
+      }
+      /* do either have a combining class of 0 (non-combining)? */
+      if (a_combining_class == 0 || b_combining_class == 0) {
+        if (combining_initialized) {
+          /* we've reached the end of the combining sequence */
+          if (combining_class_next != 0) {
+            /* prepare for the next pass */
+            utf8proc_uint8_t *elem = &(combining_classes_finished[combining_class_current/CHAR_BIT]);
+            const utf8proc_uint8_t mask = (utf8proc_uint8_t)(1 << (combining_class_current % CHAR_BIT));
+            *elem |= mask;
+            combining_class_current = combining_class_next;
+            combining_class_next = 0;
+            a->str = a_combining_start;
+            b->str = b_combining_start;
+            decomposing_current = decomposing_combining_start;
+            a_consume = true;
+            b_consume = true;
+            continue;
+          }
+          /* else exit combining mode and carry on as normal */
+          combining_initialized = false;
+        }
+        if (a_combining_class != b_combining_class) {
+          /* mismatch detected */
```

Comment:
Not sure why this is a mismatch, rather than just two characters that haven't been sorted?

Reply:
This is inside the branch for at least one of the combining classes being 0, the sorting is only relevant for nonzero combining classes to my understanding. I'll need to debug it more to be sure though.